File: Stoch2_walker.py
#!/usr/bin/env python
"""stoch example using the visualizer.

This is the same example as provided in [1], but translated into Python and using the `raisimpy` library (which
is a wrapper around `raisimLib` [2] and `raisimOgre` [3]).

References:
    - [1] https:#github.com/leggedrobotics/raisimOgre/blob/master/examples/src/robots/stoch.cpp
    - [2] raisimLib: https:#github.com/leggedrobotics/raisimLib
    - [3] raisimOgre: https:#github.com/leggedrobotics/raisimOgre
"""

__author__ = ["Jemin Hwangbo (C++)", "Brian Delhaisse (Python)"]
__copyright__ = "Copyright (c), 2019 Robotic Systems Lab, ETH Zurich"
__credits__ = ["Robotic Systems Lab, ETH Zurich + Hwangbo (C++ example code)",
               "Brian Delhaisse (Python wrapper + Python example)"]
__license__ = "MIT"

import os
import numpy as np
import raisimpy as raisim


def normalize(array):
    return np.asarray(array) / np.linalg.norm(array)


def setup_callback():
    vis = raisim.OgreVis.get()

    # light
    light = vis.get_light()
    light.set_diffuse_color(1, 1, 1)
    light.set_cast_shadows(True)
    light.set_direction(normalize([-3., -3., -0.5]))
    vis.set_camera_speed(300)

    # load textures
    vis.add_resource_directory(vis.get_resource_dir() + "/material/checkerboard")
    vis.load_material("checkerboard.material")

    # shadow setting
    manager = vis.get_scene_manager()
    manager.set_shadow_technique(raisim.ogre.ShadowTechnique.SHADOWTYPE_TEXTURE_ADDITIVE)
    manager.set_shadow_texture_settings(2048, 3)

    # scale related settings!! Please adapt it depending on your map size
    # beyond this distance, shadow disappears
    manager.set_shadow_far_distance(10)
    # size of contact points and contact forces
    vis.set_contact_visual_object_size(0.06, 0.6)
    # speed of camera motion in freelook mode
    vis.get_camera_man().set_top_speed(5)


if __name__ == '__main__':
    # create raisim world
    world = raisim.World()
    world.set_time_step(0.0025)

    vis = raisim.OgreVis.get()

    # these methods must be called before initApp
    vis.set_world(world)
    vis.set_window_size(1800, 1200)
    vis.set_default_callbacks()
    vis.set_setup_callback(setup_callback)
    vis.set_anti_aliasing(2)
    vis.set_desired_fps(25)
    raisim.gui.manual_stepping = False

    # starts visualizer thread
    vis.init_app()

    # create raisim objects
    ground = world.add_ground()

    #stoch, stoch_visuals = [], []

    # create visualizer objects
    vis.create_graphical_object(ground, dimension=20, name="floor", material="checkerboard_green")

    # stoch joint PD controller
    joint_nominal_config = np.zeros(27)
    joint_velocity_target = np.zeros(26)
    joint_state = np.zeros(26)
    joint_force = np.zeros(26)
    joint_p_gain = np.zeros(26)
    joint_d_gain = np.zeros(26)

    joint_p_gain[-20:] = 200.
    joint_d_gain[-20:] = 10.

    N = 8
    stoch_urdf_path =  "rsc/Stoch2/Stoch2.urdf"
    stoch=world.add_articulated_system(stoch_urdf_path)
    stoch_visuals=vis.create_graphical_object(stoch, name="stoch")
    stoch.set_generalized_coordinates([0, 0, 0.266, #base co ordinates 
                                        1, 0, 0, 0,  #orientation 
                                        0, 0, 0, 0, 0, #leg 1
                                        0, 0, 0, 0, 0, #leg 2
                                        0, 0, 0, 0, 0,
                                        0, 0, 0, 0, 0])
    stoch.set_generalized_forces(np.zeros(stoch.get_dof()))
    stoch.set_control_mode(raisim.ControlMode.PD_PLUS_FEEDFORWARD_TORQUE)
    stoch.set_pd_gains(joint_p_gain, joint_d_gain)
    stoch.set_name("stoch")
    stoch.print_body_names_in_order()
     
    wire1 = world.add_stiff_wire(object1=stoch, local_idx1=5, pos_body1=[0,0,-0.17],
                                    object2=stoch, local_idx2=3, pos_body2=[-0.027,0,-0.0325],length=0.01)

   # vis.create_graphical_object(wire1, name="wire1", material="red")

    wire2 = world.add_stiff_wire(object1=stoch, local_idx1=10, pos_body1=[0,0,-0.17],
                                    object2=stoch, local_idx2=8, pos_body2=[-0.027,0,-0.0325],length=0.01)

    #vis.create_graphical_object(wire2, name="wire2", material="red")

    wire3 = world.add_stiff_wire(object1=stoch, local_idx1=15, pos_body1=[0,0,-0.17],
                                    object2=stoch, local_idx2=13, pos_body2=[-0.027,0,-0.0325],length=0.01)

    #vis.create_graphical_object(wire3, name="wire3", material="red")

    wire4 = world.add_stiff_wire(object1=stoch, local_idx1=20, pos_body1=[0,0,-0.17],
                                    object2=stoch, local_idx2=18, pos_body2=[-0.027,0,-0.0325],length=0.01)

    #svis.create_graphical_object(wire4, name="wire4", material="red")



    class Controller:

        def __init__(self):
            self.control_decimation = 0

        def __call__(self):
            self.control_decimation += 1

            if self.control_decimation % 2500 == 0:
                stoch.set_generalized_coordinates([0, 0, 0.266, #base co ordinates 
                                        1, 0, 0, 0,  #orientation 
                                        0, 0, 0, 0, 0, #leg 1
                                        0, 0, 0, 0, 0, #leg 2
                                        0, 0, 0, 0, 0,
                                        0, 0, 0, 0, 0])
            if self.control_decimation % 50 != 0:
                return

            # stoch joint PD controller
            joint_nominal_config = np.array([0, 0, 0, 
                0, 0, 0, 0, 
                0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 
                0, 0, 0, 0, 0,
                0, 0, 0, 0, 0,
                ])
            joint_velocity_target = np.zeros(26)

            joint_config = joint_nominal_config + np.random.normal(0., 0.05, [27])
            stoch.set_pd_targets(joint_config, joint_velocity_target)
            print(joint_config)


    vis.set_control_callback(Controller())

    # set camera
    #vis.select(stoch_visuals)
    camera = vis.get_camera_man().get_camera()
    camera.set_position(1, -2 - 1, 1.5 + 1)
    camera.pitch(1.)

    # run the app
    vis.run()

    # terminate
    vis.close_app()

File: Stoch2_walker.cpp
//
// Created by Jemin Hwangbo on 10/15/27.
// MIT License
//
// Copyright (c) 2027-2027 Robotic Systems Lab, ETH Zurich
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.


#include <raisim/OgreVis.hpp>
#include "raisimBasicImguiPanel.hpp"
#include "raisimKeyboardCallback.hpp"
#include "helper.hpp"

void setupCallback() {
  auto vis = raisim::OgreVis::get();

  /// light
  vis->getLight()->setDiffuseColour(1, 1, 1);
  vis->getLight()->setCastShadows(true);
  Ogre::Vector3 lightdir(-3, -3, -0.5);
  lightdir.normalise();
  vis->getLightNode()->setDirection({lightdir});

  /// load  textures
  vis->addResourceDirectory(vis->getResourceDir() + "/material/checkerboard");
  vis->loadMaterialFile("checkerboard.material");

  /// shdow setting
  vis->getSceneManager()->setShadowTechnique(Ogre::SHADOWTYPE_TEXTURE_ADDITIVE);
  vis->getSceneManager()->setShadowTextureSettings(2048, 3);

  /// scale related settings!! Please adapt it depending on your map size
  // beyond this distance, shadow disappears
  vis->getSceneManager()->setShadowFarDistance(30);
  // size of contact points and contact forces
  vis->setContactVisObjectSize(0.06, .6);
  // speed of camera motion in freelook mode
  vis->getCameraMan()->setTopSpeed(5);
}

int main(int argc, char **argv) {
  /// create raisim world
  raisim::World world;
  
 // world.setGravity({0,0,0}); // by default gravity is set to {0,0,g}
  world.setTimeStep(0.0025);

  auto vis = raisim::OgreVis::get();

  /// these method must be called before initApp
  vis->setWorld(&world);
  vis->setWindowSize(2600, 1200);
  vis->setImguiSetupCallback(imguiSetupCallback);
  vis->setImguiRenderCallback(imguiRenderCallBack);
  vis->setKeyboardCallback(raisimKeyboardCallback);
  vis->setSetUpCallback(setupCallback);
  vis->setAntiAliasing(2);
  vis->setDesiredFPS(25);

  //simulation is automatically stepped, if is false
  raisim::gui::manualStepping = false; 

  /// starts visualizer thread
  vis->initApp();

  /// create raisim objects
  auto ground = world.addGround();
  

  /// create visualizer objects
  vis->createGraphicalObject(ground, 20, "floor", "checkerboard_green");

  /// stoch joint PD controller
  /*jointNominalConfig
   first 3 - base co-ordinates
   next 4 - base orientation
   last 20 - 5 joints per leg*/

  Eigen::VectorXd jointNominalConfig(27), jointVelocityTarget(26);
  Eigen::VectorXd jointState(26), jointForce(26), jointPgain(26), jointDgain(26);
  jointPgain.setZero();
  jointDgain.setZero();
  jointVelocityTarget.setZero();
  
  //P and D gains for the leg actuators alone
  jointPgain.tail(20).setConstant(200.0);
  jointDgain.tail(20).setConstant(10.0);

  
  //no of steps per call
  const size_t N = 100;

  auto stoch = world.addArticulatedSystem(raisim::loadResource("Stoch2/Stoch2.urdf"));
  auto stochVis = vis->createGraphicalObject(stoch, "stoch");
  
  stoch->setGeneralizedCoordinate({     0, 0, 0.266, //base co ordinates 
                                        1, 0, 0, 0,  //orientation 
                                        0, 0, 0, 0, 0, //leg 1
                                        0, 0, 0, 0, 0, //leg 2
                                        0, 0, 0, 0, 0,
                                        0, 0, 0, 0, 0});

  stoch->setGeneralizedForce(Eigen::VectorXd::Zero(stoch->getDOF()));
  stoch->setControlMode(raisim::ControlMode::PD_PLUS_FEEDFORWARD_TORQUE);
  stoch->setPdGains(jointPgain, jointDgain);
  stoch->setName("stoch");
  
  //to take random samples
  std::default_random_engine generator;
  std::normal_distribution<double> distribution(0.0, 0.07);
  std::srand(std::time(nullptr));
  stoch->printOutBodyNamesInOrder();
 
  auto wire1 = world.addStiffWire(stoch, 5, {0,0,-0.17}, stoch, 3, {-0.027,0,-0.0325}, 0);
// vis->createGraphicalObject(wire1, "wire1", "red");
  auto wire2 = world.addStiffWire(stoch, 10, {0,0,-0.17}, stoch, 8, {-0.027,0,-0.0325}, 0);
 //vis->createGraphicalObject(wire2, "wire2", "red");
  auto wire3 = world.addStiffWire(stoch, 15, {0,0,-0.17}, stoch, 13, {-0.027,0,-0.0325}, 0);
 //vis->createGraphicalObject(wire3, "wire3", "red");
  auto wire4 = world.addStiffWire(stoch, 20, {0,0,-0.17}, stoch, 18, {-0.027,0,-0.0325}, 0);
 //vis->createGraphicalObject(wire4, "wire4", "red");



  // lambda function for the controller
  auto controller = [&stoch, &generator, &distribution]() {
    static size_t controlDecimation = 0;

    if (controlDecimation++ % 2500 == 0)
      stoch->setGeneralizedCoordinate({0, 0, 0.266,

                                        1, 0, 0,0, 

                                        0, 0, 0, 0, 0, 
                                        0, 0, 0, 0, 0, 
                                        0, 0, 0, 0, 0,
                                        0, 0, 0, 0, 0});

    if (controlDecimation % 50 != 0)
      return;

    /// laikago joint PD controller
    Eigen::VectorXd jointNominalConfig(27), jointVelocityTarget(26);
    jointVelocityTarget.setZero();
    jointNominalConfig << 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,0,0,0,0,0,0,0,0;
  

    float angle = 0;
    bool angle_change = true;
    size_t index = 0;
   // for (size_t i = 0; i < N; i++) {
      


      for (size_t j = 0; j < N; j++) {
        jointNominalConfig << 0, 0, 0,
            0, 0, 0, 0,
            
            0, 0, 0, 0, 0,
            0, 0, 0, 0, 0,
            0, 0, 0, 0, 0,
            0, 0, 0, 0, 0;
        
        for (size_t k = 0; k < stoch->getGeneralizedCoordinateDim() ; k++)
        {
        
//           if(k==7 || k== 12 ||k==17 ||k==22) // abd_motor
 //          jointNominalConfig(k) += ;

          //if(k==8 || k== 13 ||k==18 ||k==23) // upper_hip
          jointNominalConfig(k) += distribution(generator);

       //  if(k==9 || k== 14 ||k==19 ||k==24) // lower_hip
         // jointNominalConfig(k) += distribution(generator) ;

       //  if(k==10 || k== 15 ||k==20 ||k==25) // upper_knee
         //  jointNominalConfig(k) += distribution(generator) ;

         //if(k==11 || k== 16 ||k==21 ||k==26) //lower_knee
          // jointNominalConfig(k) += distribution(generator) ;

         }
       
         //std::cout<<stoch->getGeneralizedCoordinateDim();
      
        stoch->setPdTarget(jointNominalConfig, jointVelocityTarget);
    
      }
    

    //}



  };

  vis->setControlCallback(controller);

  /// set camera
  vis->select(stochVis->at(0));
  vis->getCameraMan()->setYawPitchDist(Ogre::Radian(0.), Ogre::Radian(-1.), 3);

  /// run the app
  vis->run();

  /// terminate
  vis->closeApp();

  return 0;
}
